test/gtest/type-builder.cpp

```diff
@@ -101,6 +102,46 @@ TEST_F(TypeTest, TypeIterator) {
   EXPECT_EQ(reverse, tuple.rend());
 }
 
+TEST_F(TypeTest, PrintTypes) {
+  TypeBuilder builder(4);
+
+  Type refStructA = builder.getTempRefType(builder[0], Nullable);
+  Type refStructB = builder.getTempRefType(builder[1], Nullable);
+  Type refArrayA = builder.getTempRefType(builder[2], Nullable);
+  Type refArrayB = builder.getTempRefType(builder[3], Nullable);
+  builder[0] = Struct({Field(refArrayB, Immutable)});
+  builder[1] = Struct({Field(refStructA, Immutable)});
+  builder[2] = Array(Field(refStructB, Immutable));
+  builder[3] = Array(Field(refArrayA, Immutable));
+
+  auto result = builder.build();
+  ASSERT_TRUE(result);
+  auto built = *result;
+
+  std::vector<HeapType> structs{built[0], built[1]};
+  std::vector<HeapType> arrays{built[2], built[3]};
+
+  using ArrayPrinter = IndexedTypeNameGenerator<DefaultTypeNameGenerator>;
+  using StructPrinter = IndexedTypeNameGenerator<ArrayPrinter>;
+  StructPrinter print(structs, "struct", arrays, "array");
```

Comment:
hmm...
```cpp
AP = ITNG<Default>;
SP = ITNG<AP>;
```
It's like a chain?

What actually ends up consuming the args on this line? `IndexedTypeNameGenerator`'s constructor just forwards them, and I'm missing where a parent does so.

Is this C++ type magic necessary? :smile:

Reply:
Yep, they compose into a chain. The full type that gets instantiated is `IndexedTypeNameGenerator<IndexedTypeNameGenerator<DefaultTypeNameGenerator>>`, but of course that's too long to write out all at once. In principle you can make it as long as you want.

What actually consumes the args is the initializer for the inner `fallback` name generator: `fallback(std::forward<Args>(args)...)`. In the case of the test here, that means that `fallback` is initialized with `arrays, "array"`.

> Is this C++ type magic necessary? 😄

Not really, no 😁 I did find this useful when debugging the new type fuzzing stuff, though. I created an indexed type printer for the new temporary types that fell back to a different indexed type printer (with a different prefix) for the original types and that made it easy to see which types came from where and how they all fit together.